File: client/src/main/java/org/dreamexposure/discal/client/listeners/discord/RoleDeleteListener.java
package org.dreamexposure.discal.client.listeners.discord;

import org.dreamexposure.discal.core.database.DatabaseManager;

import discord4j.common.util.Snowflake;
import discord4j.core.event.domain.role.RoleDeleteEvent;
import reactor.core.publisher.Mono;

public class RoleDeleteListener {
    public static Mono<Void> handle(final RoleDeleteEvent event) {
        Mono<Boolean> updateRsvps = DatabaseManager.removeRSVPRole(event.getGuildId(), event.getRoleId());

        Mono<Boolean> updateControlRole = DatabaseManager.getSettings(event.getGuildId())
            .filter(settings -> !"everyone".equalsIgnoreCase(settings.getControlRole()))
            .filter(settings -> event.getRoleId().equals(Snowflake.of(settings.getControlRole())))
            .doOnNext(settings -> settings.setControlRole("everyone"))
            .flatMap(DatabaseManager::updateSettings);

        return Mono.when(updateRsvps, updateControlRole);
    }
}

File: client/src/main/kotlin/org/dreamexposure/discal/client/listeners/discord/RoleDeleteListener.kt
package org.dreamexposure.discal.client.listeners.discord

import discord4j.core.event.domain.role.RoleDeleteEvent
import org.dreamexposure.discal.core.business.GuildSettingsService
import org.dreamexposure.discal.core.business.RsvpService
import org.springframework.stereotype.Component

@Component
class RoleDeleteListener(
    private val rsvpService: RsvpService,
    private val settingsService: GuildSettingsService,
) : EventListener<RoleDeleteEvent> {

    override suspend fun handle(event: RoleDeleteEvent) {
        rsvpService.removeRoleForAll(event.guildId, event.roleId)

        val settings = settingsService.getSettings(event.guildId)
        if (settings.controlRole != null && event.roleId == settings.controlRole) {
            settingsService.upsertSettings(settings.copy(controlRole = null))
        }
    }
}
